### ISSUE TITLE
Support Optional Plugins

### ISSUE THREAD

**Issue by anthonyalayo** [ascopes/protobuf-maven-plugin#182]
Plugins like https://github.com/alpancs/protoc-gen-pubsub-schema which are helpful but not necessary for the build, are essentially optional. Could we have an option to make a plugin optional?

I trialed this locally and it worked perfectly:
```
  public boolean generate(GenerationRequest request) throws ResolutionException, IOException {
    var protocPath = discoverProtocPath(request);

    var plugins = discoverPlugins(request);
    var importPaths = discoverImportPaths(request);
    var sourcePaths = discoverCompilableSources(request);

    if (sourcePaths.isEmpty()) {
      if (request.isFailOnMissingSources()) {
        log.error("No protobuf sources found. If this is unexpected, check your "
            + "configuration and try again.");
        return false;
      } else {
        log.warn("No protobuf sources found; nothing to do!");
        return true;
      }
    }

    if (request.isPluginOnly() && plugins.isEmpty()) {
      log.warn("No plugins found; nothing to do!");
      return true;
    }
```

It's currently using my `<isPluginOnly>` attribute to make it cleaner.

**Comment by ascopes** [comment]
For my understanding, what sort of situations would this be needed in? In the example of the pubsub schema plugin, are you unable to resolve it locally?

I ask because if this is purely due to development vs CI access or configuration, a Maven profile might be more appropriate for enabling and disabling plugin functionality.

For example...

```xml
<profiles>
  <profile>
    <id>generate-pubsub-schemas</id>
    <build>
      <plugins>
        <plugin>
          <groupId>io.github.ascopes</groupId>
          <artifactId>protobuf-maven-plugin</artifactId>
          <version>${protobuf-maven-plugin.version}</version>
          <executions>
            <execution>
              <id>generate-pubsub-schema</id>
              <goals><goal>generate</goal></goals>
              <configuration>
                <binaryUrlPlugins>
                  <binaryUrlPlugin>
                    <url>http://...</url>
                  </binaryUrlPlugin>
                </binaryUrlPlugins>
              </configuration>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </build>
  </profile>
</profiles>
```

If this doesn't address your use case then I'll be happy to look into adding this but I might need a few more details on exactly why it is needed.

My main concern is ignoring missing plugins may affect build reproducibility (e.g. if you expect it to work in CI but it silently doesn't run, then you may get malformed deployments of your project).

**Comment by anthonyalayo** [comment]
@ascopes great points. At first a profile (inherited from a parent pom) wasn't working for me. I just went back to it after reading you message and it is indeed working now. I think a profile is definitely the right approach. Closing this up, thanks!

**Comment by anthonyalayo** [comment]
After some fiddling, I remembered why I moved away from a profile. Let's say I want to run just the plugin, with that particular profile. 

When I try something like this:
```
mvn protobuf:generate -P my-extra-setup
```

It's not running the executions behind the profile, it's only running the non-profiled (default java generation) part.

When I run this:
```
mvn compile -P ardi-generate
```

It works. I hit that issue, and hacked in a "ignore plugin if not found" to get around this. Any idea why?

**Comment by anthonyalayo** [comment]
After some fiddling, I remembered why I moved away from a profile. Let's say I want to run just the plugin, with that particular profile. 

When I try something like this:
```
mvn protobuf:generate -P my-extra-setup
```

It's not running the executions behind the profile, it's only running the non-profiled (default java generation) part.

When I run this:
```
mvn compile -P ardi-generate
```

It works. I hit that issue, and hacked in a "ignore plugin if not found" to get around this. Any idea why?

**Comment by anthonyalayo** [comment]
After some fiddling, I remembered why I moved away from a profile. Let's say I want to run just the plugin, with that particular profile. 

When I try something like this:
```
mvn protobuf:generate -P my-extra-setup
```

It's not running the executions behind the profile, it's only running the non-profiled (default java generation) part.

When I run this:
```
mvn compile -P my-extra-setup
```

It works. I hit that issue, and hacked in a "ignore plugin if not found" to get around this. Any idea why?

**Comment by ascopes** [comment]
ahh, have you tried adding a phase on the profile execution?

**Comment by anthonyalayo** [comment]
@ascopes I dug into this, and I believe that when executing a plugin goal directly, it won't actually invoke a profile:

![image](https://github.com/ascopes/protobuf-maven-plugin/assets/17555572/eae36a9f-2ce1-4746-89cf-a134d1d8b142)

That was a diff of the two commands above run with `-X`. With that being said, the idea of an optional plugin starts to become more necessary. I can simply run `mvn protobuf:generate -P my-extra-setup` locally when needed, and my CI environment won't fail due to extra binaries that aren't available. What do you think?

**Comment by ascopes** [comment]
Yeah, sounds good to me. I was thinking about adding an attribute to each plugin block with the v2 changes to allow  opting specific plugins in and out... would that be something of interest to you to work on?

**Comment by anthonyalayo** [comment]
Yeah sure, i'll make a PR for it. Any preference on naming?

**Comment by ascopes** [comment]
probably `skipIfMissing`, which just defaults to false.

It might actually just be worth applying it to the URL and Path plugin interfaces.

You might need to do a bit of internal plumbing to get it to work nicely though.

If at all possible it'd be cool to throw together an integration test for this as well as the unit tests are not in the best shape at the moment (unless you fancy a go at that).

Thanks :-)

**Comment by anthonyalayo** [comment]
How's this @ascopes ? https://github.com/ascopes/protobuf-maven-plugin/pull/183

**Comment by anthonyalayo** [comment]
How's this @ascopes ? https://github.com/ascopes/protobuf-maven-plugin/pull/183

I have to context switch to some other tasks, but feel free to take it to the finish line if there's anything you don't like?

**Comment by ascopes** [comment]
Cool, I can pick it up if you want me to, it might need to wait until the weekend when I am not working though so I can review it properly.

**Comment by anthonyalayo** [comment]
No worries, same boat here, just added what I could to your latest comments. Appreciate it!

**Comment by ascopes** [comment]
Released 2.0.0-alpha2 with this on it. Should be up on Maven Central for use.

**Comment by anthonyalayo** [comment]
Thanks @ascopes !

**Comment by ascopes** [comment]
Released under v2.0.0 as stable!
